// @hapiness/biim/commonjs/biim.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@hapiness/core");
const _1 = require("./");
let ReformatBadRequestLifecycle = class ReformatBadRequestLifecycle {
    onEvent(request, reply) {
        /* istanbul ignore else */
        if (request.response.isBoom && !request.response.output.payload['key']) {
            const source = request.response.output;
            const err = _1.Biim.create(source.statusCode, source.payload.error);
            request.response.output.payload['key'] = err.output.payload['key'];
        }
        reply.continue();
    }
};
ReformatBadRequestLifecycle = __decorate([
    core_1.Lifecycle({
        event: 'onPreResponse'
    })
], ReformatBadRequestLifecycle);
let BiimModule = class BiimModule {
};
BiimModule = __decorate([
    core_1.HapinessModule({
        version: '1.4.0',
        declarations: [ReformatBadRequestLifecycle],
    })
], BiimModule);
exports.BiimModule = BiimModule;
