// @hapiness/logger/commonjs/module/logger.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@hapiness/core");
const logger_service_1 = require("./logger.service");
const logger_lifecycle_1 = require("./logger.lifecycle");
const logger_config_1 = require("./logger.config");
let LoggerModule = LoggerModule_1 = class LoggerModule {
    static setConfig(config) {
        return {
            module: LoggerModule_1,
            providers: [{ provide: logger_config_1.LOGGER_CONFIG, useValue: config }]
        };
    }
};
LoggerModule = LoggerModule_1 = __decorate([
    core_1.HapinessModule({
        version: '1.1.0',
        declarations: [logger_lifecycle_1.AccessLogs],
        providers: [logger_service_1.LoggerService],
        exports: [logger_service_1.LoggerService]
    })
], LoggerModule);
exports.LoggerModule = LoggerModule;
var LoggerModule_1;
