// @hapiness/core/commonjs/core/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ExtentionHooksEnum;
(function (ExtentionHooksEnum) {
    ExtentionHooksEnum[ExtentionHooksEnum["OnExtensionLoad"] = 'onExtensionLoad'] = "OnExtensionLoad";
    ExtentionHooksEnum[ExtentionHooksEnum["OnModuleInstantiated"] = 'onModuleInstantiated'] = "OnModuleInstantiated";
})(ExtentionHooksEnum = exports.ExtentionHooksEnum || (exports.ExtentionHooksEnum = {}));
var ModuleEnum;
(function (ModuleEnum) {
    ModuleEnum[ModuleEnum["OnStart"] = 'onStart'] = "OnStart";
    ModuleEnum[ModuleEnum["OnError"] = 'onError'] = "OnError";
    ModuleEnum[ModuleEnum["OnRegister"] = 'onRegister'] = "OnRegister";
})(ModuleEnum = exports.ModuleEnum || (exports.ModuleEnum = {}));
/**
 * Represents the position where
 * the module is instantiate
 */
var ModuleLevel;
(function (ModuleLevel) {
    ModuleLevel[ModuleLevel["ROOT"] = 0] = "ROOT";
    ModuleLevel[ModuleLevel["PRIMARY"] = 1] = "PRIMARY";
    ModuleLevel[ModuleLevel["SECONDARY"] = 2] = "SECONDARY";
})(ModuleLevel = exports.ModuleLevel || (exports.ModuleLevel = {}));
